src/index.ts

```diff
@@ -1,7 +1,79 @@
+import { euclidean } from 'ml-distance-euclidean';
 /**
- * My module
- * @returns A very important number
+ * DBSCAN
+ * @returns list of cluster，-1 is noise
  */
-export function myModule(): number {
-  return 42;
+interface IPoint {
+  index: number;
+  value: number[];
+  label: number;
+}
+function rangeQuery(
+  cur: IPoint,
+  data: IPoint[],
+  epsilon: number,
+  distance: (p: number[], q: number[]) => number,
+) {
+  return data.filter((point) => distance(point.value, cur.value) <= epsilon);
+}
+export function dbscan(
+  points: number[][],
+  epsilon: number = 1,
+  minPoints: number = 2,
+  distance: (p: number[], q: number[]) => number = euclidean,
+) {
+  if (!(points instanceof Array)) {
+    throw Error('points must be of type array, ' + typeof points + ' given');
+  }
+  let clusterId = 0;
+  const data: IPoint[] = points.map((point, index) => ({
+    index,
+    value: point,
+    label: -1,
+  }));
+
+  data.forEach((point) => {
```

Comment:
When making algorithm we should avoid forEach that is quite slow (and a little bit more difficult to read). You may have a look at this kind of benchmark: https://jsbench.me/phkg20iq7z

Please use a `for (const point of data)`

Reply:
ok